[WIP] MKL repos download for mac and windows + update to MKL(-DNN) 0.12

anton-matosov opened the pull request:
This PR implements MKL repos for Mac (tested) and Windows (not tested yet)
As well as reworks MKL.BUILD to support all 3 desktop platforms

**Update 1** Unfortunately, because of bazel bug https://github.com/bazelbuild/bazel/issues/4480 you will still experience loader issues on Mac and you will be forced to install MKL globally or deal with DYLD_LIBRARY_PATH to overcome invalid rpath in generated .so files of tensorflow (similar to https://github.com/tensorflow/tensorflow/issues/6729 for CUDA libraries in dependencies).

**Update 2** Bazel 0.12.0 (not yet released) fixes the issue and it builds and runs flawlessly (verified with [0.12.0rc1](https://releases.bazel.build/0.12.0/rc1/index.html) )

anton-matosov committed:
Fix MKL build on MacOS

anton-matosov committed:
Fix invalid return type, user_memory_->set_data_handle(buf) returns void.

anton-matosov committed:
Fix type mismatch

anton-matosov committed:
Implement downloading and linking of MKL repo on mac and windows

anton-matosov committed:
Rework the way MKL libs are binded to avoid warnings and simplify code

anton-matosov committed:
Switch to matching version of MKL-DNN

anton-matosov committed:
Add mirror links

anton-matosov committed:
Add renamed license file

anton-matosov committed:
Add -fopenmp flag for linux mkl build as Slice op kernel is optimized using OpenMP

gunan commented on tensorflow/workspace.bzl:
```diff
@@ -60,6 +60,11 @@ def check_bazel_version_at_least(minimum_bazel_version):
 
+# Sanitize a dependency so that it works correctly from code that includes
+# TensorFlow as a submodule.
+def clean_dep(dep):
```
Let's either modify the whole file, or split this out of this PR as a separate PR.
Just want to make sure the file is consistent.

anton-matosov replied:
Sure. I will update all occurrences, if that is acceptable =)

anton-matosov replied:
Done

gunan reviewed (changes requested):
All build and bzl changes are OK with me except for the one comment. @andydavis1 could you take a look at the cc file changes? I am especially worried about the return type change in `mkl_util.h`, `SetUsrMemHandle`

anton-matosov commented:
@gunan `mkl_util.h` changes fixes a bug, which breaks even linux build. it was discussed here https://github.com/tensorflow/tensorflow/pull/17392#pullrequestreview-101268242

anton-matosov committed:
Use clean_dep in the entire file for consistency

anton-matosov committed:
Merge commit '70cd9ed2d2ea37a6da6f813a99b32c03e90736a4' into MKL_repos_download

# Conflicts:
#	tensorflow/workspace.bzl

anton-matosov committed:
Fix build. Constants are declrared as static in header outside the class scope now

anton-matosov committed:
Reworked fix for `-openmp` flag using the suggestion from @gunan
Original version here https://github.com/gunan/tensorflow/commit/b4d39c5226e8bbafe084b1dfc35f4644f2d906c2

anton-matosov commented:
Bazel 0.12.0 (not yet released) fixes the linker issue and it builds and runs flawlessly with downloaded MKL libs (verified with 0.12.0rc1 )

sb2nov commented:
@anton-matosov can you resolve the merge conflicts.

anton-matosov committed:
Merge commit '67e2efa6f1cb38b3d79e6c8096eb3f8f4ea9c610' into MKL_repos_download

# Conflicts:
#	tensorflow/workspace.bzl

anton-matosov commented:
@sb2nov done

anton-matosov commented:
🎉

dfumento commented:
According to Update 2 Bazel 0.12.0 (not yet released) fixes the issue and it builds and runs flawlessly (verified with 0.12.0rc1 )

I tried with the install version (not homebrew) of 0.12.0 bazel  using this line:

bazel build --config=opt --config=mkl //tensorflow/tools/pip_package:build_pip_package

and I receive this message:
error: unknown target CPU 'armv7-a' 

There are messages on-line that suggest that this is a problem with 0.12.0, but 0.12.0 is needed for the MKL support.

anton-matosov commented:
bazel 0.12.0 was release couple days ago, check it out (I didn't yet, but should work)
to bypass CPU 'armv7-a' issue with pre-release binary specify `--cpu=darwin` argument  (for mac build) to the bazel call. More details here https://github.com/tensorflow/tensorflow/pull/17392#issuecomment-375420834
